File: go/allergies/allergies.go
package allergies

var allergens = []string{
	"eggs",
	"peanuts",
	"shellfish",
	"strawberries",
	"tomatoes",
	"chocolate",
	"pollen",
	"cats",
}

// Allergies returns the allergies that a person with a given allergy score
// is allergic to.
func Allergies(score int) []string {
	result := make([]string, 0)
	for i, allergen := range allergens {
		if score&(1<<uint(i)) != 0 {
			result = append(result, allergen)
		}
	}
	return result
}

// AllergicTo checks whether a given allergy score implies an allergy to
// a given allergy.
func AllergicTo(score int, allergen string) bool {
	for i, a := range allergens {
		if a == allergen {
			return score&(1<<uint(i)) != 0
		}
	}
	return false
}

File: haskell/allergies/Allergies.hs
module Allergies where

import Data.Bits ((.&.), unsafeShiftL)

data Allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats
              deriving (Enum, Eq, Show)

type Score = Int

allergies :: Score -> [Allergen]
allergies score = filter (`isAllergicTo` score) [Eggs ..]

isAllergicTo :: Allergen -> Score -> Bool
isAllergicTo allergen score = score .&. (1 `unsafeShiftL` fromEnum allergen) > 0
